File: benchmark/autolifter/dac/longest_odd10s.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = 0;
Config SampleIntMax = 1;
Config SampleSize = 20;

Inductive List = cons {Int, List} | nil Unit;

is_even = \a: Int. == a (* 2 (/ a 2));

length = fix (\f: List->Int. \x: List.
  match x with
    cons {h, t} -> + (f t) 1
  | nil _ -> 0
  end
);

concat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {h, t} -> cons {h, f t y}
  | nil _ -> y
  end
);

split = \xs: List. fix (\f: List->Int->{List, List}. \x: List. \n: Int.
  if (< n 1) then {nil unit, x}
  else match x with
    cons {h, t} -> let res = f t (- n 1) in {cons {h, res.1}, res.2}
  | _ -> {nil unit, nil unit}
  end
) xs (/ (length xs) 2);

dac = \v: List -> Int. \xs: List.
  let run = fix (\f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {_, nil _} -> xs
    | _ -> let sp = split xs in
      concat (f sp.1) (f sp.2)
    end
  ) in v (run xs);

inf = 100;

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

longest_odd10s = fix (
  \f: Bool -> Bool -> Int -> List -> Int. \s1: Bool. \s2: Bool.
  \len: Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} ->
    let s1 = and s2 (== h 1) in
    let s2 = == h 0 in
    let len = if s1 then + 1 len else if s2 then len else 0 in
    if is_even len then f s1 s2 len t
    else max len (f s1 s2 len t)
  end
) false false 0;

main = dac longest_odd10s;

File: ml-benchmark/autolifter/dac/longest_odd10s.ml
type list =
  | Cons of int * list
  | Nil

let is_even a = a = 2 * (a / 2)

let rec length x =
  match x with
  | Cons (_, t) -> length t + 1
  | Nil -> 0

let rec concat x y =
  match x with
  | Cons (h, t) -> Cons (h, concat t y)
  | Nil -> y

let split xs =
  let rec f x n =
    if n < 1 then (Nil, x)
    else
      match x with
      | Cons (h, t) ->
        let res = f t (n - 1) in
        match res with
        | (a, b) -> (Cons (h, a), b)
      | _ -> (Nil, Nil)
  in
  f xs (length xs / 2)

let dac v =
  fun xs ->
    let rec run xs =
      match xs with
      | Nil -> xs
      | Cons (_, t) ->
        match t with
        | Nil -> xs
        | _ ->
          let sp = split xs in
          match sp with
          | (a, b) -> concat (run a) (run b)
    in
    v (run xs)

let inf = 100

let max a b = if a < b then b else a

let rec longest_odd10s_f s1 s2 len xs =
  match xs with
  | Nil -> 0
  | Cons (h, t) ->
    let s1 = s2 && h = 1 in
    let s2 = h = 0 in
    let len = if s1 then len + 1 else if s2 then len else 0 in
    if is_even len then longest_odd10s_f s1 s2 len t
    else max len (longest_odd10s_f s1 s2 len t)

let longest_odd10s = longest_odd10s_f false false 0

let program = dac longest_odd10s